staging/src/k8s.io/apiserver/pkg/server/genericapiserver_graceful_termination_test.go

```diff
@@ -429,58 +464,82 @@ func TestMuxAndDiscoveryComplete(t *testing.T) {
 }
 
 func TestPreShutdownHooks(t *testing.T) {
-	s := newGenericAPIServer(t, true)
-	doer := setupDoer(t, s.SecureServingInfo)
-
-	preShutdownHookErrCh := make(chan error)
-	err := s.AddPreShutdownHook("test-backend", func() error {
-		// this pre-shutdown hook waits for the requests in flight to drain
-		// and then send a series of requests to the apiserver, and
-		// we expect these series of requests to be completed successfully
-		<-s.lifecycleSignals.InFlightRequestsDrained.Signaled()
-
-		// we send 5 requests, once every second
-		var r result
-		client := newClient(true)
-		for i := 0; i < 5; i++ {
-			r = doer.Do(client, func(httptrace.GotConnInfo) {}, fmt.Sprintf("/echo?message=attempt-%d", i), 1*time.Second)
-			if r.err != nil {
-				break
+	tests := []struct {
+		name   string
+		server func() *GenericAPIServer
+	}{
+		{
+			name: "ShutdownSendRetryAfter is disbaled",
+			server: func() *GenericAPIServer {
+				return newGenericAPIServer(t, false)
+			},
+		},
+		{
+			name: "ShutdownSendRetryAfter is enabled",
+			server: func() *GenericAPIServer {
+				return newGenericAPIServer(t, true)
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			s := test.server()
+			doer := setupDoer(t, s.SecureServingInfo)
+
+			preShutdownHookErrCh := make(chan error)
+			err := s.AddPreShutdownHook("test-backend", func() error {
+				// this pre-shutdown hook waits for the shutdown duration to elapse,
+				// and then send a series of requests to the apiserver, and
+				// we expect these series of requests to be completed successfully
+				<-s.lifecycleSignals.AfterShutdownDelayDuration.Signaled()
```

Comment:
have changed it from `<-s.lifecycleSignals.InFlightRequestsDrained.Signaled()` to `<-s.lifecycleSignals.AfterShutdownDelayDuration.Signaled()`, otherwise it would deadlock with the changes introduced by https://github.com/kubernetes/kubernetes/pull/110000

Reply:
Which seems correct to me (not just because of deadlock, but also conceptually...)